Kafka, MongoDB, Redis: Publish templated connectors (#324)
mulesoft/docs-connectors: modules/ROOT/pages/kafka/kafka-connector-studio.adoc

= Kafka Studio Configuration - Mule 4
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../../assets/images/

To configure a connector in Anypoint Studio:

. Install the connector.
. Configure the connector.
. Configure an input source for the connector.

== Install the Connector

From within Studio:

. In Studio, create a Mule Project.
. In the Mule Palette, click *(X) Search in Exchange*.
. In *Add Modules to Project*, type the name of the connector in the search field.
. Click the connector name in *Available modules*.

> . Click *Add >*.

Reviewer: Is the UI Add >
?
Or is that a typo?

Author: In the UI:
![AddModulesToProject](https://user-images.githubusercontent.com/8439645/63456179-32ef4b80-c403-11e9-96d4-6416bf8bd231.png)

. Click *Finish*.

> Or using Exchange:

Reviewer: Did Kim OK the formatting of this? I would make these two diff procedures H2

== Install the Connector in Studio
== Install the Connector Using Exchange

Or similar

. In Studio, create a Mule Project.
. Click the Exchange *(X)* icon in the Studio task bar.
. In Exchange, click *Login* and supply your Anypoint Platform username and password.
. In Exchange, select *All assets* and search for "Kafka".
. Select Kafka, click *Add to project*.
. Follow the prompts to install the connector.

== Configure the Connector

. Drag and drop the connector to the Studio Canvas.
. To create a global element for the connector, set these fields:
+
.. Basic:

> ** Bootstap Servers - Comma-separated host-port pairs used for establishing the initial connection to the Kafka cluster. This is the same as the "bootstrap.servers" value you must provide to Kafka clients

Reviewer: ```suggestion
** Bootstrap Servers - Comma-separated host-port pairs used for establishing the initial connection to the Kafka cluster. This is the same as the `bootstrap.servers` value you must provide to Kafka clients
```

(producer or consumer).
** Additional properties - Additional properties as key and value that you need for your connection. Here you can put whatever property Kafka supports.
+
image::kafka/kafka-basic-studio-config.png[Basic Configuration]
+
.. SSL:
** All the parameters from Basic configuration.

> ** Key Store Type - The file format of the key store file. This is optional and default value is "JKS".

Reviewer: ```suggestion
** Key Store Type - The file format of the key store file. This is optional and the default value is `JKS`.
```

> ** Key Store Password - The store password for the key store file. This is optional and only needed if "keyStoreLocation" is configured.

Reviewer: This is another repeat comment, so this is the last time I'll comment. You can S/R for the others.

> ** Key Store Password - The store password for the key store file. This is optional and only needed if "keyStoreLocation" is configured.

Reviewer: ```suggestion
** Key Store Password - The store password for the key store file. This is optional and needed only if `keyStoreLocation` is configured.
```

** Key Store Location - The location of the key store file. This is optional and can be used for two-way authentication for connector.
** Trust Store Type - The file format of the trust store file.
** Trust Store Password - The password for the trust store file.
** Trust Store Location - The location of the trust store file.
+
image::kafka/kafka-ssl-studio-config.png[SSL Configuration]
+
.. Kerberos:
** All the parameters from Basic configuration.
** Principal - Kerberos principal

> ** Keytab - Path to keytab file associated with "principal".

Reviewer: This has been all over the place ("principal"). I don't know what it is or why it's in double quotes. If it's a property or value, it should prob be `principal` but since I don't know I haven't commented.

> ** Service Name - The Kerberos principal name that Kafka Broker runs as.

Reviewer: ```suggestion
** Service Name - The Kerberos principal name that Kafka broker runs as.
```
S/R Kafka Broker and change to Kafka broker

> ** Additional JAAS Properties - Additional properties as key->value that you need to set on "sasl.jaas.config" and that you usually include in JAAS configuration file.

Reviewer: key-value pairs
`sasl.jaas.config`

+
image::kafka/kafka-kerberos-studio-config.png[Kerberos Configuration]
+
.. Kerberos SSL:
** All the parameters from Basic configuration.
** All the parameters from SSL configuration.
** All the parameters from Kerberos configuration.
+
image::kafka/kafka-kerberos-ssl-studio-config.png[Kerberos SSL Configuration]
+

> . Based on the operation that you have dragged on the canvas, configure the following fields:

Reviewer: ```suggestion
. Based on the operation that you dragged on the canvas, configure the below fields:
```

.. Consumer trigger:

> ** Topic - name of Kafka topic to consume messages from.

Reviewer: In most of the lists, it is init capped for the definition. For example:
Topic - Name of Kafka topic...

I think init cap is the correct way to go and we should be consistent. I'm not going to comment on all the instances of this.

> ** Partition offsets (Optional) - list of Offset representing partitions offsets configuration. For each element in the list you have to specify partition index and offset.

Reviewer: ```suggestion
** Partition offsets (Optional) - list of offsets representing partitions offsets configuration. For each element in the list you have to specify partition index and offset.
```

I have the usual problem with this one... I don't know what it's saying.

+
image::kafka/kafka-consumer-studio-config.png[Consumer Studio Configuration]
+
.. Producer operation:

> ** Topic - Topic to send the message to

Reviewer: ```suggestion
** Topic - Topic to send the message to.
```

> ** Key - Key belonging to the message that is going to be sent

Reviewer: ```suggestion
** Key - Key belonging to the message to send.
```

> ** Message - Message to be sent

Reviewer: ```suggestion
** Message - Message to send.
```

+
image::kafka/kafka-producer-studio-config.png[Producer Studio Configuration]

== Configure an Input Source

Configure an input source for the connector such as a connector operation, using an *HTTP Listener*, or *Scheduler*.

The *Message Consumer* operation can be used as an input source in your connector.

Parameters:

[%header,cols="30s,70a"]
|===
| Name | Description
| Configuration |The name of the configuration to use.
| Topic |Name of the topic to consume messages from.

> | Partition Offsets a| Array of Offset - List of Offset representing partitions offsets configuration. For each element in the list you have to specify partition index and offset.

Reviewer: Offsets ? Plural ?  Usually a list of something implies there are multiple.

> | Primary Node Only |Whether this source should only be executed on the primary node when running in Cluster. 

Reviewer: ```suggestion
| Primary Node Only |Whether this source should be executed only on the primary node when running in a Cluster. 
```
